kaska/smoothn.py

```diff
@@ -1,537 +1,617 @@
+        s_max_bnd = None
+
+    return s_min_bnd, s_max_bnd
+
+
+def bound_calc(h_limit, rank):
+    """
+    Calculate smooth bound from leverage bound.
+    (h/rnk)**2 = (1 + a)/( 2 a)
+    a = 1/(2 (h/rnk)**2 -1)
+    where a = sqrt(1 + 16 s)
+    (a**2 -1)/16
+    """
+    return np.sqrt((
+                ((1 + np.sqrt(1 + 8 * h_limit ** (2.0 / rank))
+                  ) / 4. / h_limit ** (2.0 / rank)
+                ) ** 2 - 1) / 16.)
+
+
+def initial_z(y, z0, is_weighted):
+    """
+    An initial conditions for z.
+    With weighted/missing data, an initial guess is provided to ensure faster
+    convergence. For that purpose, a nearest neighbor interpolation followed
+    by a coarse smoothing are performed.
+
+    For unweighted data, the initial guess is zero.
+    """
+    if is_weighted:
+        if z0 is not None:  # an initial guess (z0) has been provided
+            z = z0
+        else:
+            z = np.where(np.isfinite(y), y, 0.0)  # initial_guess(y,IsFinite);
+    else:
+        z = np.zeros_like(y)
+
+    return z
+
+
+def init_xpost(s, s_min_bnd, s_max_bnd, is_auto):
+    """
+    Calculate xpost based on the smoothing and smoothing bounds values.
+    """
+    if is_auto:
+        try:
+            return np.array([(0.9 * np.log10(s_min_bnd) +
+                              np.log10(s_max_bnd) * 0.1)])
+        except ValueError:
+            return np.array([100.0])
     else:
-        RobustIterativeProcess = False; # stop the whole process
-
-  ## Warning messages
-  #---
-  if isauto:
-    if abs(log10(s)-log10(sMinBnd))<errp:
-        warning('MATLAB:smoothn:SLowerBound',\
-            ['s = %.3f '%(s) + ': the lower bound for s '\
-            + 'has been reached. Put s as an input variable if required.'])
-    elif abs(log10(s)-log10(sMaxBnd))<errp:
-        warning('MATLAB:smoothn:SUpperBound',\
-            ['s = %.3f '%(s) + ': the upper bound for s '\
-            + 'has been reached. Put s as an input variable if required.'])
-    #warning('MATLAB:smoothn:MaxIter',\
-    #    ['Maximum number of iterations (%d'%(MaxIter) + ') has '\
-    #    + 'been exceeded. Increase MaxIter option or decrease TolZ value.'])
-  return z,s,exitflag,Wtot
-
-def warning(s1,s2):
-  print(s1)
-  print(s2[0])
-
-## GCV score
-#---
-#function GCVscore = gcv(p)
-def gcv(p,Lambda,aow,DCTy,IsFinite,Wtot,y,nof,noe,smoothOrder):
-    # Search the smoothing parameter s that minimizes the GCV score
-    #---
-    s = 10**p;
-    Gamma = 1./(1+(s*abs(Lambda))**smoothOrder);
-    #--- RSS = Residual sum-of-squares
-    if aow>0.9: # aow = 1 means that all of the data are equally weighted
+        return np.array([np.log10(s)])
+
+
+#  GCV score
+# ---
+# function GCVscore = gcv(p)
+def gcv(p, lambda_v, aow, dct_y, is_finite, w_tot, y, nof, noe, smooth_order):
+    """
+    Search the smoothing parameter s that minimizes the GCV score
+    """
+    s = 10 ** p
+    gamma = 1.0 / (1 + (s * np.abs(lambda_v)) ** smooth_order)
```

Comment:
gamma is calculated in one more place than here - maybe move to its own function/lambda?

Reply:
Done